docs/repo/debugging/async-test-hangs.md

```diff
@@ -0,0 +1,270 @@
+# Debugging Async Test Hangs
+
+(Adapted from an internal document originally written by @jaredpar)
+
+Tests hanging on CI machines are often caused by async tests that are blocked waiting for results that will never complete. Unlike traditional synchronous hangs, attaching to an async hang and looking at executing threads typically will not provide anything useful as there won't be any threads actually executing the "awaiting"  methods.
+
+This document walks through debugging a test hang using WinDBG and SOS.
+
+
+## Prerequisites 
+
+- [WinDBG](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools)
+- Basic understanding of async machinery  
+
+## Loading SOS
+
+Open WinDBG and attach to the hanging process or open the dump, then to load sos:
+
+
+32-bit dump/process:
+
+```
+> .load C:\Windows\Microsoft.NET\Framework\v4.0.30319\sos.dll
+```
+
+64-bit dump/process:
+
+```
+> .load C:\Windows\Microsoft.NET\Framework64\v4.0.30319\sos.dll
+```
+
+## Determining the executing test
+
+We need to find the xUnit types in memory that track test exception, in this case; [InvokeTestAsync](https://github.com/xunit/xunit/blob/9d10262a3694bb099ddd783d735aba2a7aac759d/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestRunner.cs#L67). It is an async method, so we're going to find the blocked tests by looking through instances of its async state machine.
+
+```
+> !dumpheap -type InvokeTestAsync
+```
+
+```
+ Address       MT     Size
+036c8cb0 08389c3c       44     
+037a1828 08389c3c       44     
+039d9c40 08389c3c       44     
+03adb7c4 08389c3c       44     
+03b9cf68 08389c3c       44     
+03bd7638 08389c3c       44     
+03bf5748 08389c3c       44     
+03d75740 08389c3c       44     
+03d7fee0 08389c3c       44     
+03dc019c 08389c3c       44     
+03dc2e9c 08389c3c       44     
+03dd51a4 08389c3c       44     
+03ee8c00 08389c3c       44     
+03f98a38 08389c3c       44     
+03ff07e4 08389c3c       44     
+0bff4e9c 08389c3c       44     
+0c037b98 08389c3c       44     
+0c05c2b0 08389c3c       44     
+
+Statistics:
+      MT    Count    TotalSize Class Name
+08389c3c       18          792 Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4
+Total 18 objects
+```
+
+The output tells me there's 18 instances of the async state machine in memory. What we are looking for are state machines that have not completed.
```

Comment:
I'm sure there must be a foreach command to dump all of these and find the ones in the right state.

When looking for all uncollected exceptions in a dump, I often use something like this, which could probably be adapted here:

`.foreach (ex {!dumpheap -type Exception -short}){.echo "********************************";!pe ${ex} }`

Reply:
I timeboxed this and gave up trying to write something here - you can use `.if` works but couldn't work a syntax that worked.

In the end, I think this doc is going to be temporary, the CPS one is much better at finding this:

![image](https://user-images.githubusercontent.com/1103906/45397586-0de74a00-b684-11e8-86a5-ec6fe9a54c8c.png)
 
Started a conversation to figure out how to make this extension public.